feat: implement the VSCode Language Model API.

julesmons opened the pull request:
This feature enables usage of language models provided by extensions within cline!
(This means: including, but not limited to, copilot!)

The implementation is still a bit rough around the edges (some todo comments), but i got it working in my vscode! I was able to generate some code using cline and my existing copilot subscription's claude 3.5 model.

I have tried to maintain the code-style as best as possible.

Any and all feedback welcome!

julesmons commented:
This should also close https://github.com/cline/cline/pull/765

julesmons commented on src/api/providers/vscode-lm.ts:
```diff
@@ -0,0 +1,313 @@
+
+                if (chunk instanceof vscode.LanguageModelToolCallPart) {
+
+                    // TODO: Determine how to handle tool calls.
```
This todo does not mean that tool calling does not work.
When i tested this implementation cline could run my terminal commands just fine!

If i understand it correctly, the VSCode language model is able to call new tools. That's what this return type seems to be for. I think VSCode uses this for the "Copilot Edits" feature. But it seems to be reduntant within cline.

I'm thinking i might add some form of error toast, or nicely pass this called tool to cline or something.

julesmons replied:
This code snippet has since been edited.

julesmons commented:
Also; This implementation should be thoroughly tested before releasing.

julesmons commented:
![Example screenshot](https://github.com/user-attachments/assets/ed096875-db87-4b4a-b76a-a3b9a5e0bdf0)

julesmons commented:
![{77CC1E2C-E81C-4C59-B107-277C01BF9395}](https://github.com/user-attachments/assets/984588e1-137c-46d9-aaf5-6a27cef3ce3e)

saoudrizwan commented:
This is very cool, I have been meaning to look into making this possible. Thank you for putting this pull request together @julesmons! I will review and test on my end when I get a chance

julesmons commented:
## Update
In order to emit token usage events to cline, the provider has to call a function of the VSCode LM API to count the tokens in a message. However, this seems to be sort of slow and i make use of it twice in every createMessage call. This seems to be slowing things down quite a lot.

I'm currently working on a fix for this.

julesmons commented:
## Update
I've committed some fixes.
These should further optimize and speed up the provider by a significant amount.

---

The following changes have been made:
1. Fewer calls to the countTokens of the model
2. Added a sliding token count cache (this is still a proposal and i need to test and make sure it works as intended)
  - Ideally, I'd like to store the count in the message history, instead of keeping a cache.
  - Currently this cache holds a sliding window of the token count of 1000 messages. This is an arbitrary value that I'd like to make configurable once i have verified this is a significant performance improvement.
3. The cancellation logic was not working as indended, which i have fixed.

julesmons commented:
## Update 3
After some testing i decided to opt for the storage of the token count in the message history.
This enables smoother operation, even after VSCode is restarted.

The sliding-window token-count cache has been removed.
However; i did have to make some changes in the `src/core/Cline.ts` file.

amaanalihaider commented:
Thank you. I am very excited about this. When It will be merged?

julesmons commented:
You're welcome! Not sure when it'll be merged though.

I've been using the provider this week and kept an eye on debug logs i've placed etc. I've realised that the `getModel` function the ApiHandler providers HAS to be async to better support the VSCode Language Model API. The provider currently passes incorrect model info to cline, the workaround i've implemented for this does not work because of the lifetime of the provider.

In addition to this, the message history data-structure has to be extended with token info (to avoid needless performance hungry token counting, because the vscode lm api does not expose usage)

Changing the core abstractions means this PR will have to become a breaking change.

julesmons commented:
It's very nice to hear all this positive feedback, much appreciated!

First of all; I've been using this provider with copilot for about two weeks now.
I wouldn't exactly call it 'stable', as it freezes occasionally and has some other minor issues.
My VSCode currently runs a custom-built version of the experimental branch over at the repo this PR is linked to.
The provider has already allowed me to greatly increase my productivity, generating in the millions of tokens per day (eventually hitting the copilot token-based rate-limit).

The experimental and development branches differ from the actual pull request feature-branch.
These branches have implementations of the breaking changes I've been talking about, in order to further improve the provider.
If you're indeed eager to try it out, you can build either the development (cline-compatible) or experimental (and unstable) branch of my forked repo.

I've also seen RooCline and think their concept/idea is awesome.
Whilst working on this provider (and in the Cline code-base in general), I've found a lot of things that I would implement differently or would like to improve upon (mainly just for the fun of it).
I want to clarify that I'm not saying the Cline code is bad. Quite the opposite actually.
This fantastic project has inspired me to experiment with several Cline-subsystems (which, of course, resulted in a slew of breaking changes).

Opening lots of pull-requests would become very time consuming and cumbersome very quickly.
This inspired me to create a fork of my own called... Recline! (pun intended)
I've renamed my forked repo and have already started work on the aforementioned experimental branch. (which currently has some animation/ui-based issues if i remember correctly)

After working on that experimental branch i decided to broaden the scope.
As I'm writing this I'm in the midst of heavy comprehensive refactoring, re-structuring, and the implementation of some new features i thought would be nice.

Some changes I've made include, but are not limited to:

- Removed 'bloat' files. For example; the webview index.html which wasn't even used.
- Switched to preact (5kb and react compatible, so massive performance boost with almost no downside)
- Removed the react-scripts template build setup
- Removed the "install:all" script and switched to pnpm-based monorepo structure.
- Changed the project to use Parcel as a bundler causing it to now have just one build step. (This also quite significantly reduced the bundle size)
- Stripped 'overkill'/'bloat' or otherwise (in my eyes) suboptimal dependencies (like cheerio, execa, globby, etc.).
  - Example: replaced globby with "fd" (which is simmilar to ripgrep but for gnu find, it's fast and automatically respects the workspaceRoot gitignore!)

I've decided to publish Recline when i have the project stable enough to try out.
The first commit to the Recline branch will probably already diverge the project too far from the original cline main branch.
This will cause auto-merging of upstream changes to be nearly impossible, forcing any clever new improvements to be implemented by hand.
I want to re-iterate that I'm doing this for fun and that I'm not sure I'll have the time to properly maintain it.
However, it will of course be open-source and any input or contributions are always welcome.

---

!!! This does not mean I'm abandoning this PR though !!!

When I've perfected the VSCode LM provider in my fork, I'll make sure to update this PR with a 'clean and cline-compatible' version of it!
I would just like to explore what _could_ be changed (what works and what does not) before deciding what _has_ to be changed/implemented, making sure no unnecessary breaking changes get merged into Cline.

In the meantime, I'm really keen to hear what @saoudrizwan thinks about the aforementioned storing the token count in the message history.
The VSCode Language Model API does not supply usage information, but has a countTokens method.
Sadly, i've found that method to be very performance heavy; exponentially slowing the provider down as the message history elongates.

Storing the count would enable the provider to only count incomming messages.
This way all messages are counted just once.

---

Cheers!

Bigsy commented:
Not to derail this PR too much but just wanted to say thanks @julesmons - I built the development branch of Recline (loving the name btw) and tried out a few advent of codes with copilot/sonnet, and it worked brilliantly. Seems to go down rabbit holes/get stuck in loops a little easier than normal sonnet but still works great. I'm guessing that's maybe the shorter context length via copilot and hope they make that bigger when its not in preview, like they did with o4.

Looking forward to a full release in whatever guise it is but will continue to use this in the meantime its really great congrats!

julesmons commented:
You're welcome! What you're experiencing has to do with the getModel info not being async.
This limitation forces the provider to return the openai sane defaults.

This means Cline has no idea what the model can actually handle...

---

The context window sure is a limitation indeed.
I've found myself using multiple models to complete a task. (the provided gpt and o1-mini models also work; Altough sonnet is miles ahead of the competition)

I'm keen to see how Microsoft further integrates sonnet into copilot.

worldnine commented:
このコミットはリリースされてますか？

julesmons commented:
I saw that a version of the vscode lm provider without the token-counting/usage reporting has been merged on main. Cool!

I'm going to close this PR as it has become redundant.

saoudrizwan commented:
Just wanted to jump in here and thank you big time for the contribution @julesmons -- I ended up pulling bits and pieces from various forks, most predicated off your dive in the VS Code LM API. Incredible work, and hope you can be involved in improving this implementation as it is very experimental (but seems to work great so far!)
